pyomo/contrib/solver/common/util.py

```diff
@@ -0,0 +1,128 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2024
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from pyomo.common.errors import PyomoException
+from pyomo.core.expr.visitor import ExpressionValueVisitor, nonpyomo_leaf_types
+import pyomo.core.expr as EXPR
+from pyomo.core.base.objective import Objective
+
+
+class NoFeasibleSolutionError(PyomoException):
+    def __init__(self):
+        super().__init__(
+            'A feasible solution was not found, so no solution can be loaded. '
+            'Please set opt.config.load_solutions=False and check '
+            'results.solution_status and '
+            'results.incumbent_objective before loading a solution.'
+        )
+
+
+class NoOptimalSolutionError(PyomoException):
+    def __init__(self):
+        super().__init__(
+            'Solver did not find the optimal solution. Set '
+            'opt.config.raise_exception_on_nonoptimal_result = False to bypass this error.'
+        )
+
+
+class NoSolutionError(PyomoException):
+    def __init__(self):
+        super().__init__(
+            'Solution loader does not currently have a valid solution. Please '
+            'check results.termination_condition and/or results.solution_status.'
+        )
+
+
+class NoDualsError(PyomoException):
+    def __init__(self):
+        super().__init__(
+            'Solver does not currently have valid duals. Please '
+            'check results.termination_condition and/or results.solution_status.'
+        )
+
+
+class NoReducedCostsError(PyomoException):
+    def __init__(self):
+        super().__init__(
+            'Solver does not currently have valid reduced costs. Please '
+            'check results.termination_condition and/or results.solution_status.'
+        )
+
+
+class IncompatibleModelError(PyomoException):
+    def __init__(self):
+        super().__init__(
+            'Model is not compatible with the chosen solver. Please check '
+            'the model and solver.'
+        )
+
+
+def get_objective(block):
+    """
+    Get current active objective on a block. If there is more than one active,
+    return an error.
+    """
+    objective = None
+    for obj in block.component_data_objects(
+        Objective, descend_into=True, active=True, sort=True
+    ):
+        if objective is not None:
+            raise ValueError('Multiple active objectives found.')
+        objective = obj
+    return objective
+
+
+class _VarAndNamedExprCollector(ExpressionValueVisitor):
```

Comment:
TODO (later): this should be ported to the `StreamBasedExpressionVisitor`

Reply:
I have lots of these conversions to do...